### PR TITLE
ClassValue substitution: Fix computeValue usage

### LINKED ISSUE #2421: Substitution for ClassValue#get cannot handle null
**Describe the issue**
I am trying to build a service as a native image that utilizes Apache WSS4J. WSS4J's `org.apache.wss4j.dom.util.WSSecurityUtil` implements a `java.lang.ClassValue<Method>` with an implementation of `java.lang.ClassValue#computeValue` that may return `null`. This causes an NPE in GraalVM's substitute `Target_java_lang_ClassValue`, which stores the result of `computeValue` in a `ConcurrentHashMap`, which does not allow `null` as key or value.

**Describe GraalVM and your environment:**
 - GraalVM version: GraalVM Version 20.0.0 CE
 - JDK major version: 11
 - OS: Linux Mint 19.3 Tricia
 - Architecture: AMD64

**More details**
WSS4J's implementation of ClassValue<Method>:
```
    private static final ClassValue<Method> GET_DOM_ELEMENTS_METHODS = new ClassValue<Method>() {
        @Override
        protected Method computeValue(Class<?> type) {
            try {
                return getMethod(type, "getDomElement");
            } catch (NoSuchMethodException e) {
                //best effort to try, do nothing if NoSuchMethodException
                return null;
            }
        }
    };
```
Substitution of `ClassValue#get` in GraalVM

```
    @Substitute
    private Object get(Class<?> type) {
        Object result = values.get(type);
        if (result == null) {
            Object newValue = computeValue(type);
            Object oldValue = values.putIfAbsent(type, newValue);
            result = oldValue != null ? oldValue : newValue;
        }
        return result;
    }
```
`values` is of type `ConcurrentHashMap`. Executing the WSS4J code for a type like `com.sun.org.apache.xerces.internal.dom.ElementNSImpl` yields an NPE:
```
java.lang.NullPointerException
	at java.util.concurrent.ConcurrentHashMap.putVal(ConcurrentHashMap.java:1011)
	at java.util.concurrent.ConcurrentHashMap.putIfAbsent(ConcurrentHashMap.java:1541)
	at java.lang.ClassValue.get(JavaLangSubstitutions.java:509)
	at org.apache.wss4j.dom.util.WSSecurityUtil.getDomElement(WSSecurityUtil.java:699)
	at org.apache.wss4j.dom.util.WSSecurityUtil.prependChildElement(WSSecurityUtil.java:370)
	at org.apache.wss4j.dom.util.WSSecurityUtil.findWsseSecurityHeaderBlock(WSSecurityUtil.java:497)
	at org.apache.wss4j.dom.message.WSSecHeader.insertSecurityHeader(WSSecHeader.java:167)
...
```

### PR DESCRIPTION
**Pull request by tfaller**
This PR fixes #2421.

The "get" method of ClassValue  has the special property that racing threads must return the same value.
In the substitution this is ensured by using a ConcurrentHashMap. The problem is, that this map can't handle null values - but null is a valid response. Without the PR a NPE occurs if null is a response.

To support null we can use a dummy null. This dummy null can be store in the map.
The dummy null is also here to distinguish between "value was computed = null" and "value was not yet computed".
Because a simple `if (newValue == null) return null;` is not correct. 
If we would do this simple check, instead of using a dummy null, the next time the "get" method is called for the same type, "computeValue" would be called again. This is not correct.

### CODE REVIEW

**Comment by tfaller** [comment]
@graalvmbot  I recently signed the OCA. I'm already listed in the public OCA list.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/JavaLangSubstitutions.java [hunk 1]
```diff
@@ -569,6 +569,8 @@ private static double log1p(double x) {
     @RecomputeFieldValue(kind = RecomputeFieldValue.Kind.Custom, declClass = JavaLangSubstitutions.ClassValueInitializer.class)//
     private final ConcurrentMap<Class<?>, Object> values;
 
+    private static final Object dummyNull = new Object();
```
[the remark this answers is not in the document]
Comment (tfaller) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/JavaLangSubstitutions.java]: @christianwimmer Thank you. Yes, `NULL_MARKER` is better. I changed it.

**Comment by jaikiran** [comment]
I think this also fixes https://github.com/oracle/graal/issues/2734

**Comment by sendev1** [comment]
Hi @christianwimmer , I have this issue https://github.com/oracle/graal/issues/2734 and I need this fix to move forward with my work. I unzipped the svm.jar and replaced the class Target_java_lang_ClassValue with code fix and tried running native image but still I am seeing NPE. To make sure the code fix is used when I run native-image, I added System.out.println in the `Target_java_lang_ClassValue.get` method but that is not getting printed. Could you please advice me how to get logs printed from Target_java_lang_ClassValue in the Windows SDK 7.1 command prompt so that I can confirm the code fix is used. More info on what ever I tried is present here also https://github.com/spring-projects-experimental/spring-graalvm-native/issues/219. Thank you.

**Comment by tfaller** [comment]
My PR definitely fixes #2421. But I'm not sure if this also fixes #2734. I actually don't think so.
@sendev1 My PR would fix your issue, if the stack trace of your NPE contains 
`at java.lang.ClassValue.get(JavaLangSubstitutions.java:509)`. But it does not, your issue has another cause.

**Comment by sendev1** [comment]
@tfaller Thanks for your reply. I am trying to replace the class Target_java_lang_ClassValue directly inside the svm.jar with your fix  but for some reason I am not seeing that the replaced class is taking effect when I run native-image. Any idea how can I try native-image with your code fix?

**Comment by christianwimmer** [comment]
@tfaller Thanks for the PR, will be merged soon

@sendev1 I added a fix for your #2734 which will be merged at the same time. Basically the null values also need to be replaced with the `NULL_MARKER` for values that are already computed at image build time.

**Comment by sendev1** [comment]
@christianwimmer could you please advise. Were there any additional code changes done to fix https://github.com/oracle/graal/issues/2734? I am not seeing any additional changes on PR https://github.com/oracle/graal/pull/2728 that was done to fix my issue. Thank you.

**Pull request merged**